\documentclass[12pt]{article}

\usepackage[utf8]{inputenc} % declare encoding as utf8
\usepackage{graphicx} % to enable \includegraphics
\usepackage{flafter} % to make sure floats appear after their position in text
\usepackage{pdflscape} % allow some pages to be in landscape orientation
\usepackage{layout} % for debugging purposes, looking at page layout
\usepackage[margin=1in]{geometry} % leaves more space for text body
\usepackage[hidelinks]{hyperref} % hyperlinks from list of figures/tables
\usepackage[labelfont=bf]{caption} % format captions and make links go to figure instead of caption
\usepackage{csvsimple} % read and display csv files as tables
\usepackage{booktabs} % prettier tables
\usepackage{threeparttable} % easy addition of footnotes below tables
\usepackage{textcomp} % makes the \textdegree command available
\usepackage{tocloft} % formatting the lists of tables and figures
\usepackage{xcolor} % enable text coloring
\usepackage{calc} % makes \widthof available to set the width of the last column
\usepackage{array} % using command \arraybackslash to restore the definition of \\ after \centering
\usepackage{longtable} % allow tables to span more than one page
\usepackage{threeparttablex} % threeparttable for longtable
\usepackage{capt-of} % put caption on next page
\usepackage{listings}

% Formatting the table of contents/subsection
\renewcommand{\contentsname}{Supplemental methods}

% Formatting the list of tables
\renewcommand{\listtablename}{Supplemental tables} % Write "Supplemental tables" instead of "List of tables"
\renewcommand\cfttabpresnum{Table } % Write "Table" before table number in list of tables
\setlength{\cfttabnumwidth}{6em} % Spacing before captions
\setlength{\cftbeforefigskip}{1ex} % Vertical spacing between entries

% Formatting the list of figures
\renewcommand{\listfigurename}{Supplemental figures} % Write "Supplemental figures" instead of "List of figures"
\renewcommand\cftfigpresnum{Figure } % Write "Figure" before fig number in list of figures
\setlength{\cftfignumwidth}{6em} % Spacing before captions
\setlength{\cftbeforetabskip}{1ex} % Vertical spacing between entries

\graphicspath{ {../figures/} } % tell Latex to look for figures in figures/

% Format figure and table numbers to follow S1, S2, ...
\renewcommand\thefigure{S\arabic{figure}} 
\renewcommand\thetable{S\arabic{table}} 

% Defining new environments as I will always want the figures centered
\newenvironment{cfigure}
	{\begin{figure} \centering}
	{\end{figure}}

\newenvironment{lsfigure}
	{\begin{landscape} \begin{figure} \centering}
	{\end{figure} \end{landscape}}

\setcounter{totalnumber}{1} % Allow a maximum of one float (table/figure) per page

\setlength{\parskip}{2ex}

% Defining a command to generate a table of phenotypic data for a single gwas
\newcommand{\gwastableone}[1]{
\begin{table}
	\caption{Numerical coding of phenotypic data used for the GWAS analysis of #1}
	\label{#1_gwas_table}

	\centering
	\begin{threeparttable}\small

		\csvreader[head to column names,
		/csv/separator=semicolon,
		tabular = llc,
		table head = \toprule Phenotype & Frequency & GWAS \#1 \\ \midrule,
		table foot = \bottomrule]%
		{../tables/#1_gwas_table.csv}%
		{}%
		{\Value & \Frequency & \GWASa}

			   \begin{tablenotes}
			   \item[a] test
			   \end{tablenotes}
	\end{threeparttable}

\end{table}
\clearpage
}

% Defining a command to generate a table of phenotypic data for two gwas
\newcommand{\gwastabletwo}[1]{
\begin{table}
	\caption{Numerical coding of phenotypic data used for the GWAS analysis of #1}
	\label{#1_gwas_table}

	\centering
	\begin{threeparttable}\small

		\csvreader[head to column names,
		/csv/separator=semicolon,
		tabular = llcc,
		table head = \toprule Phenotype & Frequency & GWAS \#1 & GWAS \# 2\\ \midrule,
		table foot = \bottomrule]%
		{../tables/#1_gwas_table.csv}%
		{}%
		{\Value & \Frequency & \GWASa & \GWASb}

			   \begin{tablenotes}
			   \item[a] test
			   \end{tablenotes}
	\end{threeparttable}

\end{table}
\clearpage
}

% Defining a command to generate a table of phenotypic data for three gwas
\newcommand{\gwastablethree}[1]{
\begin{table}
	\caption{Numerical coding of phenotypic data used for the GWAS analysis of #1}
	\label{#1_gwas_table}

	\centering
	\begin{threeparttable}\small

		\csvreader[head to column names,
		/csv/separator=semicolon,
		tabular = llccc,
		table head = \toprule Phenotype & Frequency & GWAS \#1 & GWAS \# 2 & GWAS \# 3\\ \midrule,
		table foot = \bottomrule]%
		{../tables/#1_gwas_table.csv}%
		{}%
		{\Value & \Frequency & \GWASa & \GWASb & \GWASc}

			   \begin{tablenotes}
			   \item[a] test
			   \end{tablenotes}
	\end{threeparttable}

\end{table}
\clearpage
}

% Defining a command for manhattan plot figures
\newcommand{\manhattanplot}[1]{
\begin{cfigure}
	\includegraphics[width = 6in]{#1_manhattan}
	\caption{#1\_manhattan}
	\label{#1_manhattan}
\end{cfigure}

\clearpage}

% Defining a new command for signal plot figures
\newcommand{\signalplot}[1]{
\begin{cfigure}
	\includegraphics[width = 6in]{#1_signal}
	\caption{#1\_signal}
	\label{#1_signal}
\end{cfigure}

\clearpage}

% Defining a new command for gene plot figures
\newcommand{\geneplot}[1]{
\begin{cfigure}
	\includegraphics[width = 6in]{#1_gene}
	\caption{#1\_gene}
	\label{#1_gene}
\end{cfigure}

\clearpage}

% Defining a new command for k-mer plot figures
\newcommand{\kmerplot}[1]{
	\begin{cfigure}
		\includegraphics[width = 6.5in]{#1_kmers}
		\caption{#1\_kmers}
		\label{#1_kmers}
	\end{cfigure}

\clearpage}

% Defining a new command for LD plot figures
\newcommand{\ldplot}[1]{
	\begin{cfigure}
		\includegraphics[width = 6.5in]{#1_ld}
		\caption{#1\_ld}
		\label{#1_ld}
	\end{cfigure}

\clearpage}

% Allows underscores in captions but is not a great solution
\catcode`_=12

% Information for the title page
\title{Supplemental data to ``ARTICLE TITLE''}
\author{Marc-André Lemay \and François Belzile}
\date{}

% END OF PREAMBLE

\begin{document}

\maketitle \thispagestyle{empty}

\tableofcontents

\vspace{4ex}

\listoftables

\thispagestyle{empty}
%\clearpage

\vspace{4ex}

\listoffigures

%\thispagestyle{empty}

\clearpage

\begin{landscape} \footnotesize
	\begin{ThreePartTable}

		\begin{TableNotes}
		\item[a] test
		\end{TableNotes}

		\begin{longtable}{llcclccccl}

			\caption{Signals}\\
			\label{signals_table}\\

			\toprule
			Trait & Locus & Chromosome & -log10(p) & Gene & Platypus & vg & Paragraph & k-mers  & Study \\
			\midrule \endfirsthead

			\caption[]{Signals \emph{(continued)}}\\

			\toprule
			Trait & Locus & Chromosome & -log10(p) & Gene & Platypus & vg & Paragraph & k-mers  & Study \\
			\midrule
			\endhead

			\bottomrule
			Continued on next page \\
			\endfoot

			\bottomrule \insertTableNotes
			\endlastfoot
			\csvreader[head to column names,
			late after line=\\,
			late after last line=,
			before reading={\catcode`\#=12},
			after reading={\catcode`\#=6},
			/csv/separator=semicolon]%
			{../tables/signals_table.csv}{}{\Trait & \Locus & \Chromosome & \Pvalues & \Gene & \Platypus & \Vg & \Paragraph & \Kmers & \Study}

		\end{longtable}

	\end{ThreePartTable}

\end{landscape}
\clearpage

\gwastableone{FLOWER.COLOR}
\gwastabletwo{PUBESCENCE.COLOR}
\gwastabletwo{SEED.COAT.COLOR}
\gwastabletwo{STEM.TERMINATION.TYPE}
\gwastablethree{HILUM.COLOR}
\gwastabletwo{POD.COLOR}
\gwastabletwo{PUBESCENCE.FORM}
\gwastableone{PUBESCENCE.DENSITY}
\gwastablethree{SEED.COAT.LUSTER}
\gwastableone{MATURITY.GROUP}

\manhattanplot{flower_color}
\ldplot{flower_color}
%\manhattanplot{flower_color_scaffolds}
\signalplot{flower_color_W1}
\geneplot{flower_color_W1}
\kmerplot{flower_color_W1}
%\signalplot{flower_color_L1}
%\geneplot{flower_color_L1}


\manhattanplot{pubescence_color_all}
\manhattanplot{pubescence_color_all_scaffolds}
\ldplot{pubescence_color_all}
%\signalplot{pubescence_color_all_Td}
\signalplot{pubescence_color_all_T}
\geneplot{pubescence_color_all_T}
\kmerplot{pubescence_color_all_T}
\signalplot{pubescence_color_all_pcGm16}

\manhattanplot{pubescence_color_nogray}
%\manhattanplot{pubescence_color_nogray_scaffolds}
\signalplot{pubescence_color_nogray_Td}
\geneplot{pubescence_color_nogray_Td}
\kmerplot{pubescence_color_nogray_Td}
%\signalplot{pubescence_color_nogray_T}
%\geneplot{pubescence_color_nogray_T}


\manhattanplot{seed_coat_color_all}
%\manhattanplot{seed_coat_color_all_scaffolds}
%\signalplot{seed_coat_color_all_G}
%\geneplot{seed_coat_color_all_G}
%\signalplot{seed_coat_color_all_T}
%\geneplot{seed_coat_color_all_T}
%\signalplot{seed_coat_color_all_O}
\signalplot{seed_coat_color_all_I}
\geneplot{seed_coat_color_all_I}
%\signalplot{seed_coat_color_all_R}
%\geneplot{seed_coat_color_all_R}

\manhattanplot{seed_coat_color_greenyellow}
%\manhattanplot{seed_coat_color_greenyellow_scaffolds}
\signalplot{seed_coat_color_greenyellow_G}
\geneplot{seed_coat_color_greenyellow_G}
\kmerplot{seed_coat_color_greenyellow_G}
%\signalplot{seed_coat_color_greenyellow_T}
%\geneplot{seed_coat_color_greenyellow_T}
%\signalplot{seed_coat_color_greenyellow_O}
%\signalplot{seed_coat_color_greenyellow_I}
%\geneplot{seed_coat_color_greenyellow_I}
%\signalplot{seed_coat_color_greenyellow_R}
%\geneplot{seed_coat_color_greenyellow_R}


\manhattanplot{stem_termination_all}
%\manhattanplot{stem_termination_all_scaffolds}
%\signalplot{stem_termination_all_Dt2}
%\geneplot{stem_termination_all_Dt2}
\signalplot{stem_termination_all_Dt1}
\geneplot{stem_termination_all_Dt1}
\kmerplot{stem_termination_all_Dt1}
%\signalplot{stem_termination_all_E3}
%\geneplot{stem_termination_all_E3}

\manhattanplot{stem_termination_sn}
\signalplot{stem_termination_sn_stGm11}
%\manhattanplot{stem_termination_sn_scaffolds}
%\signalplot{stem_termination_sn_Dt2}
%\geneplot{stem_termination_sn_Dt2}
%\signalplot{stem_termination_sn_Dt1}
%\geneplot{stem_termination_sn_Dt1}
%\signalplot{stem_termination_sn_E3}
%\geneplot{stem_termination_sn_E3}


\manhattanplot{hilum_color_all}
\manhattanplot{hilum_color_all_scaffolds}
\signalplot{hilum_color_all_T}
\geneplot{hilum_color_all_T}
\kmerplot{hilum_color_all_T}
%\signalplot{hilum_color_all_O}
\signalplot{hilum_color_all_I}
\geneplot{hilum_color_all_I}
%\kmerplot{hilum_color_all_I}
%\signalplot{hilum_color_all_R}
%\geneplot{hilum_color_all_R}
%\signalplot{hilum_color_all_W1}
%\geneplot{hilum_color_all_W1}

\manhattanplot{hilum_color_blackbrown}
%\manhattanplot{hilum_color_blackbrown_scaffolds}
%\signalplot{hilum_color_blackbrown_T}
%\geneplot{hilum_color_blackbrown_T}
%\signalplot{hilum_color_blackbrown_O}
%\signalplot{hilum_color_blackbrown_I}
%\geneplot{hilum_color_blackbrown_I}
\signalplot{hilum_color_blackbrown_R}
\geneplot{hilum_color_blackbrown_R}
%\signalplot{hilum_color_blackbrown_W1}
%\geneplot{hilum_color_blackbrown_W1}

\manhattanplot{hilum_color_rbr}
%\manhattanplot{hilum_color_rbr_scaffolds}
%\signalplot{hilum_color_rbr_T}
%\geneplot{hilum_color_rbr_T}
%\signalplot{hilum_color_rbr_O}
%\signalplot{hilum_color_rbr_I}
%\geneplot{hilum_color_rbr_I}
%\signalplot{hilum_color_rbr_R}
%\geneplot{hilum_color_rbr_R}
%\signalplot{hilum_color_rbr_W1}
%\geneplot{hilum_color_rbr_W1}


\manhattanplot{pod_color_all}
%\manhattanplot{pod_color_all_scaffolds}
\signalplot{pod_color_all_L1}
\geneplot{pod_color_all_L1}
\signalplot{pod_color_all_L2}

\manhattanplot{pod_color_blbr}
\ldplot{pod_color_blbr}
\signalplot{pod_color_blbr_pdcGm15}
%\manhattanplot{pod_color_blbr_scaffolds}
%\signalplot{pod_color_blbr_L2}
%\signalplot{pod_color_blbr_L1}
%\geneplot{pod_color_blbr_L1}


\manhattanplot{pubescence_form_all}
%\manhattanplot{pubescence_form_all_scaffolds}
\signalplot{pubescence_form_all_Pa1}
\geneplot{pubescence_form_all_Pa1}
\kmerplot{pubescence_form_all_Pa1}
\signalplot{pubescence_form_all_Pa2}
\signalplot{pubescence_form_all_pfGm04}
\signalplot{pubescence_form_all_pfGm15}

\manhattanplot{pubescence_form_noerect}
%\manhattanplot{pubescence_form_noerect_scaffolds}
\signalplot{pubescence_form_noerect_Pa1}
\geneplot{pubescence_form_noerect_Pa1}
%\signalplot{pubescence_form_noerect_Pa2}


\manhattanplot{pubescence_density}
%\manhattanplot{pubescence_density_scaffolds}
%\signalplot{pubescence_density_Pd1}
\signalplot{pubescence_density_Ps}
\geneplot{pubescence_density_Ps}
%\signalplot{pubescence_density_P1}
%\geneplot{pubescence_density_P1}


\manhattanplot{seed_coat_luster_all}
%\manhattanplot{seed_coat_luster_all_scaffolds}
%\signalplot{seed_coat_luster_all_I}
%\geneplot{seed_coat_luster_all_I}
%\signalplot{seed_coat_luster_all_Bq}
%\signalplot{seed_coat_luster_all_B1}
%\signalplot{seed_coat_luster_all_Hps}
%\geneplot{seed_coat_luster_all_Hps}
\signalplot{seed_coat_luster_all_B}
\geneplot{seed_coat_luster_all_B}
\signalplot{seed_coat_luster_all_sclGm09}

\manhattanplot{seed_coat_luster_nointermediate}
%\manhattanplot{seed_coat_luster_nointermediate_scaffolds}
%\signalplot{seed_coat_luster_nointermediate_I}
%\geneplot{seed_coat_luster_nointermediate_I}
%\signalplot{seed_coat_luster_nointermediate_Bq}
%\signalplot{seed_coat_luster_nointermediate_B1}
%\signalplot{seed_coat_luster_nointermediate_Hps}
%\geneplot{seed_coat_luster_nointermediate_Hps}
%\signalplot{seed_coat_luster_nointermediate_B}
%\geneplot{seed_coat_luster_nointermediate_B}

\manhattanplot{seed_coat_luster_dullshiny}
%\manhattanplot{seed_coat_luster_dullshiny_scaffolds}
%\signalplot{seed_coat_luster_dullshiny_I}
%\geneplot{seed_coat_luster_dullshiny_I}
%\signalplot{seed_coat_luster_dullshiny_Bq}
%\signalplot{seed_coat_luster_dullshiny_B1}
%\signalplot{seed_coat_luster_dullshiny_Hps}
%\geneplot{seed_coat_luster_dullshiny_Hps}
%\signalplot{seed_coat_luster_dullshiny_B}
\geneplot{seed_coat_luster_dullshiny_B}
\signalplot{seed_coat_luster_dullshiny_sclGm20}

\manhattanplot{corrected_dry_weight}
\manhattanplot{corrected_dry_weight_scaffolds}
\signalplot{corrected_dry_weight_cdwGm15}
\geneplot{corrected_dry_weight_cdwGm15}


\manhattanplot{maturity_group}
%\manhattanplot{maturity_group_scaffolds}
%\signalplot{maturity_group_E1}
%\geneplot{maturity_group_E1}
%\signalplot{maturity_group_E2}
%\geneplot{maturity_group_E2}
%\signalplot{maturity_group_E3}
%\geneplot{maturity_group_E3}
%\signalplot{maturity_group_E4}
%\geneplot{maturity_group_E4}

\manhattanplot{oil}
%\manhattanplot{oil_scaffolds}
\signalplot{oil_oilGm15}
\geneplot{oil_oilGm15}
%\signalplot{oil_oilGm20}

\manhattanplot{protein}
%\manhattanplot{protein_scaffolds}
%\signalplot{protein_proteinGm15}
%\signalplot{protein_proteinGm20}

\end{document}

