\chapter*{Supplemental data}
\label{supdata}

% Format figure and table numbers to follow S1, S2, ...
\renewcommand\thefigure{S\arabic{figure}} 
\setcounter{figure}{0}
\setcounter{table}{0}
\renewcommand\thetable{S\arabic{table}} 

% Reading the contents of the variables.txt file first
\DTLloaddb[noheader, keys={key,value}]{variables}{variables.txt}
% Defining a new command to fetch the value from the key
\newcommand{\getvar}[1]{\DTLfetch{variables}{key}{#1}{value}}

% Detailed methods for some aspects of the paper
\section*{Detailed methods}
\label{annexe-sv-gwas-detailed-methods}

\subsection*{Discovery of structural variation from Illumina data}
\label{sv-gwas-sv-discovery-illumina}

We used the Illumina whole-genome sequencing (WGS) data from the 389 selected \emph{G. max} accessions
for SV discovery using methods similar to those described by
\cite{lemay2022}.  Briefly, we used four different SV discovery programs or
pipelines: 

\begin{enumerate}

\item We used AsmVar \citep[version of 2015-04-16, ][]{liu2015} to call SVs
	from \emph{de novo} assemblies generated with SOAPdenovo2 v. 2.04
		\citep{luo2012} and aligned to the reference genome using LAST
		v. 1047 \citep{kielbasa2011}.
\item We ran Manta v. 1.6.0 \citep{chen2016} in randomly selected batches: 77 batches of
	5 samples and 1 batch of 4 samples. We used the candidate SVs
		identified from each of the batches for downstream analyses.
\item We used smoove v. 0.2.4 \citep{pedersen2019} to obtain a dataset of candidate SVs
	using the following sequence of commands:
	\texttt{smoove call}, \texttt{smoove merge}, \texttt{smoove
		genotype} and \texttt{smoove paste}.
		For three (3) of the 389 samples, we were unable
		to obtain candidate variants due to a reproducible
		error (segmentation fault) when running \texttt{smoove
		call}.
	\item We used SvABA v. 1.1.3 \citep{wala2018} to call SVs from each of the
	samples separately. We classified the SVs called as breakends by SvABA
		as either deletions, duplications or inversions, and converted
		them to explicit sequence using a custom R script. Variants
		called as indels by SvABA were used as is for downstream
		analyses.
\end{enumerate}

We filtered the SVs called by each tool to remove those meeting any of the following conditions:

\begin{itemize}
	\item smaller than 50 bp or larger than 500 kb in size
	\item located on unanchored scaffolds or organellar genomes
	\item classified as unresolved breakends (\texttt{SVTYPE=BND})
	\item containing at least one N in either the reference or alternate allele sequence
\end{itemize}

Next, we converted all filtered variants to a sequence-explicit format using
\texttt{bayesTyperTools convertAllele} \citep{sibbesen2018} and normalized them
using bcftools \texttt{norm} v. 1.10.2-105 \citep{li2009-samtools}.

A single VCF file was generated from the variants called using Illumina data by
running SVmerge \citep{wong2010} in two steps. In the first step, we merged
very similar variants among the ones called by the same tool using a
sliding window of 5 bp. Next, we used the output of the first merging steps
for all four tools and merged similar variants across the four datasets using
a sliding window of 15 bp. This approach was used because a
sliding window of 15 bp on the outputs of all tools without pre-merging
resulted in a computationally intractable number of pairwise comparisons.

\subsection*{Discovery of structural variation from Oxford Nanopore sequencing data}
\label{sv-gwas-sv-discovery-ont}

We used SVs discovered among 17 Canadian soybean cultivars sequenced using
Oxford Nano\-pore Technologies (ONT) by \cite{lemay2022} as an additional
source of candidate SVs. Although this set of cultivars does not overlap the
ones analyzed in this study, it is likely that many of the SVs occurring
in Canadian germplasm also occur in the germplasm used for this study given that modern soybean
cultivars are derived from a rather small set of progenitors. We merged the SVs detected in the
long-read (ONT) data with the ones discovered in the short-read (Illumina) data
using SVmerge and systematically favored SVs discovered by Illumina whenever
SVs from the two sets were to be merged.  Our rationale for favoring variants
discovered from the Illumina data was that whenever both technologies discover
the same variants, the breakpoints of SVs discovered from Illumina data are
likely to be more precise given the higher basecalling accuracy.

\subsection*{Discovery of structural variation from high-quality assemblies}
\label{sv-gwas-sv-discovery-assembly}

In addition to Illumina data from 389 accessions, we also used
high-quality genome assemblies of 29 \emph{G. max} or \emph{Glycine soja}
accessions to call SVs. We
retrieved 26 high-quality genomes assembled by \cite{liu2020-pangenome} as well
as the genome of the cultivar Zhonghuang 13 (ZH13) from the Genome Warehouse
repository (\url{https://ngdc.cncb.ac.cn/gwh/}). Furthermore, we retrieved the genomes
of \emph{G. soja} accession W05 and \emph{G. max} cultivar Lee from SoyBase
\citep{grant2010}.

We used methods based on \cite{liu2020-pangenome} to call SVs from the
genome assemblies.  First, we used the nucmer program of the MUMmer suite of
tools v. 4.0.0rc1 \citep{marcais2018} with option \texttt{-c 1000} to align
each of the assemblies to the Williams82 assembly. Next, we used the
delta-filter program of the MUMmer suite to keep only one-to-one alignments
between any genome and the Williams82 reference using the \texttt{-1} option.
The filtered alignments were used to call SVs with a version of the svmu
program \citep{chakraborty2019} that we forked from the original repository
(\url{https://github.com/malemay/svmu}) in order to improve execution time and
memory usage.

We used a set of tools that we assembled in an R package called svmutools
\url{https://github.com/malemay/svmutools} to convert the svmu output to VCF
format. We limited ourselves to processing deletions and insertions because of
the complexity of resolving other types of variants from whole-genome
assemblies.  Copy number variants and inversions called by svmu were therefore
excluded from downstream analyses. The
resulting set of variants was filtered to remove those smaller than 50 bp or larger than 500
kb and those with any N in either the reference or alternate allele sequence.
Finally, we normalized the VCF files using \texttt{bcftools norm} and merged
them using SVmerge to obtain a single file with variants discovered through
comparison of high-quality genome assemblies.

\subsection*{Genotyping structural variation using Paragraph}
\label{sv-gwas-sv-genotyping-paragraph}

We used Paragraph v. 2.4a \citep{chen2019}
to generate a set of SV genotypes from the Illumina data of the 389 samples used in
GWAS analyses.  The set of SVs used as input for Paragraph was generated using SVmerge by
merging variants discovered from the Illumina and ONT data with
those discovered from the high-quality assemblies and svmu.
Variants in this candidate set were prepared for input to Paragraph by first removing
those closer than 150 bp (the maximum read length in our dataset) to any
chromosome end and then padding the variant representation using a script written by
\cite{hickey2020}. The latter step was necessary for the variants to be used
as input to Paragraph.

We ran Paragraph individually on each sample as recommended by the authors.
We computed the average sequencing depth for each sample from the output of the
\texttt{samtools coverage} command and used 20 times that value for the
\texttt{-M} parameter of Paragraph.  We ran Paragraph using the
\texttt{multigrmpy} command and merged the resulting VCF files using
\texttt{bcftools merge} in order to get a  single genotype file for the whole
population. Finally, we filtered the genotyped SVs by setting the genotype calls made
from less than two reads to missing and by removing variants with a minor allele
frequency < 0.02 or a proportion of missing data > 0.5. The resulting dataset
of 186,306 called SVs was used for GWAS.

\subsection*{Comparison of WGS and SoySNP50K genotypes}
\label{annexe-sv-gwas-soysnp50k}
When we first analyzed the results from the GWAS analysis on all 389 samples,
we found discrepancies between the haplotypes at some loci and the expected
phenotype that should correspond to that haplotype for simple traits such as
flower color and seed coat color. Upon further analysis, we found that a
mismatch between the sequencing data and the identity of the accession was the
most probable cause of these discrepancies. In order to identify the samples
for which this could be the case, we compared the genotypes of SNPs obtained
from the SoySNP50K array \cite{song2013} to those predicted from the Illumina WGS
data. We did this for the 385 samples that had a match in the GRIN database. Briefly, we
retrieved SoySNP50K genotype calls from SoyBase \citep{grant2010} and extracted
all SNPs with a MAF $\geq$ 0.1 in our population. We then translated the positions
of these SNPs from assembly version 2 to assembly version 4 of Williams by
finding exact matches of 41-bp sequences surrounding the SNP positions. The
32,852 SNPs for which unambiguous positions could be found on genome assembly
version 4 were genotyped from the mapped WGS reads using \texttt{bcftools
mpileup} and \texttt{bcftools call}. We then used \texttt{bcftools gtcheck} to
compute the concordance between the WGS genotype calls and the
SoySNP50K genotype dataset. We identified 24 samples
with < 90\% concordance between the two genotype datasets and removed them from
the dataset. We also excluded an additional sample for which concordance was over 90\%
but poor raw sequencing data quality (\%GC content) suggested contamination.
The concordance between WGS data and SoySNP50K genotypes is included in
Additional file 2 and shown in Figure \ref{concordance-histogram}.

\subsection*{Analysis of extremely low \textit{p}-values}
\label{annexe-sv-gwas-pvalues}

For one trait (flower color), the significance of some \emph{k}-mers was so
high that their \emph{p}-values were numerically equal to 0. In order to
compute $-\log_{10} (p)$ for visualization in Manhattan plots in these cases,
we set the \emph{p}-value to the smallest numerical value that could be
represented using R on our machine (2.225074e-308). In this particular case,
the values represented on the Manhattan plots represent an upper bound on the
\emph{p}-value and not an exact value.

% Detailed results that are not meant to fit the main paper
\section*{Detailed results}
\label{annexe-sv-gwas-detailed-results}

This section provides a more detailed per-trait account of some of the results obtained
by the GWAS analyses. Aspects of the results that were
already discussed in the main text are not repeated here; in such cases,
readers are simply redirected to specific tables and figures containing
detailed results from all four GWAS approaches. A summary of the signals found
at all loci using all four methods is presented in Table \ref{signals_table}.

\subsection*{Flower color}
\label{sv-gwas-flower-color}

We conducted a GWAS analysis contrasting purple and white flower colors
(Table \ref{FLOWER.COLOR_gwas_table}). Of the two classical loci reported by
\cite{bandillo2017}, \textit{W1} was detected by all four GWAS approaches whereas
\textit{L1} was detected by none of the approaches (Figure
\ref{flower_color_manhattan}). These results are unsurprising as W1 is the
major contributor to flower color in soybean, whereas the \textit{L1} locus
(typically associated with pod color) was detected with a much less significant
\textit{p}-value (see Table \ref{signals_table}) and only conjectured as
associated with flower color by \cite{bandillo2017}. Detailed results for the
\textit{W1} locus can be viewed in figures \ref{flower-color-main-figure},
\ref{flower_color_W1_signal} and \ref{flower_color_W1_gene}.

The \textit{k}-mer GWAS analysis also identified presumably spurious significant
\textit{k}-mers scattered throughout the genome (Figure
\ref{flower_color_manhattan}). An analysis of the pairwise LD between a subset
of those \textit{k}-mers showed that most \textit{k}-mers formed a single LD block
linked to the \textit{W1} locus (Figure \ref{flower_color_ld}). A small LD block
occurred on chromosome Gm09, however it was in moderate LD with the one on chromosome
Gm13 and therefore probably does not represent a separate locus.

\subsection*{Pubescence color}
\label{sv-gwas-pubescence-color}

We conducted two different GWAS analyses on pubescence color. The first analysis
contrasted gray, tawny and light tawny pubescence colors (Table
\ref{PUBESCENCE.COLOR_gwas_table}) and detected a signal overlapping the
\emph{T} locus using all four approaches (Figure
\ref{pubescence_color_all_manhattan}). Detailed results at the \textit{T} locus
can be viewed in figures \ref{pubescence_color_all_T_signal},
\ref{pubescence_color_all_T_gene} and \ref{pubescence_color_all_T_kmers}. We
observed that not all accessions with gray pubescence harbored the deleted
nucleotide, which could indicate either that other causal variants are resulting in
a non-functional F3’H in this population, or that the causal
variant lies elsewhere.

This first GWAS analysis also revealed a second locus on chromosome Gm16
(Figure \ref{pubescence_color_all_manhattan}). Upon closer analysis, this locus
appeared to be in strong LD with \emph{k}-mers mapping to chromosome Gm01 and
most importantly in moderate LD with the \emph{T} locus (Figure
\ref{pubescence_color_all_ld}), which suggests that this does not represent an
independent locus controlling pubescence color. This interpretation is also
supported by the fact that the signal at that locus only spans about 10 kb on
chromosome Gm16.

In the presence of a functional allele at the \emph{T} locus, the genotype at
the \emph{Td} locus influences the exact shade of pubescence color.  Therefore,
we conducted a second GWAS analysis contrasting accessions with tawny and light
tawny pubescence in order to ignore the epistatic effect of the \emph{T} locus
and instead focus on the effect of the \emph{Td} locus on pubescence color
(Table \ref{PUBESCENCE.COLOR_gwas_table}). Detailed results at this locus
can be viewed in figures \ref{pubescence-color-main-figure}, \ref{pubescence_color_nogray_Td_signal} and
\ref{pubescence_color_nogray_Td_gene}). 

\subsection*{Seed coat color}
\label{sv-gwas-seed-coat-color}

We conducted two different GWAS analyses on seed coat color. The first GWAS
included five different phenotypes observed in the dataset (Table
\ref{SEED.COAT.COLOR_gwas_table}) and resulted in the detection of strong
signals overlapping the \emph{I} locus by all four approaches (Figure
\ref{seed_coat_color_all_manhattan}). All approaches but vg also detected
signals at the \emph{G} locus. In addition, the \emph{k}-mer approach also
detected weak signals overlapping the \emph{T} and \emph{O} loci. However, the
\emph{T} and \emph{O} loci are likely chance associations given the low \textit{p}-value
associated with the \emph{T} locus and the proximity of the \emph{O} locus to
the \emph{I} locus. Detailed results of this first GWAS and the \textit{I}
locus can be viewed in figures \ref{seed_coat_color_all_manhattan} and
\ref{seed_coat_color_all_I_signal}.

The second GWAS analysis on seed coat color focused on the subset of accessions
with yellow or green seed coat only (Table \ref{SEED.COAT.COLOR_gwas_table}),
as this difference is controlled by the \emph{G} locus. Detailed results of
this analysis can be viewed in figures \ref{seed-coat-color-main-figure},
\ref{seed_coat_color_greenyellow_manhattan},
\ref{seed_coat_color_greenyellow_G_signal} and
\ref{seed_coat_color_greenyellow_G_gene}.

\subsection*{Stem termination type}
\label{sv-gwas-stem-termination}

We conducted two different GWAS analyses on stem termination type. The first GWAS
included all phenotypes observed in the dataset (determinate, semi-determinate,
and indeterminate; Table \ref{STEM.TERMINATION.TYPE_gwas_table}). The results
of this analysis at the \textit{Dt1} locus can be viewed in figures \ref{stem_termination_all_manhattan},
\ref{stem_termination_all_Dt1_signal}, \ref{stem_termination_all_Dt1_gene} and
\ref{stem_termination_all_Dt1_kmers}.

The second GWAS analysis included only
semi-determinate and indeterminate phenotypes. We performed this analysis in an attempt to also detect signals
at the \emph{Dt2} locus (Table \ref{STEM.TERMINATION.TYPE_gwas_table}).
However, this was unsuccessful using all four GWAS
approaches (Figure \ref{stem_termination_sn_manhattan}). Similarly, we failed to
detect signals at the \textit{E3} locus, which is more commonly associated
with maturity but was mentioned as associated with stem termination type by
\cite{bandillo2017}.

\subsection*{Hilum color}
\label{sv-gwas-hilum-color}

We conducted three different GWAS analyses on hilum color. The first GWAS
included seven colors (Table \ref{HILUM.COLOR_gwas_table}) and detected signals
at the \emph{I}, \emph{T} and \emph{R} loci using all four approaches except vg
at the \emph{R} locus (Figure \ref{hilum_color_all_manhattan}). No signals
were detected at the \emph{W1} and \emph{O} loci studied by
\cite{bandillo2017}. The detailed results for the \textit{I} locus can be viewed
in figure \ref{hilum_color_all_I_signal} whereas those for the \textit{T} locus
can be viewed in figures \ref{hilum_color_all_T_signal}, \ref{hilum_color_all_T_gene}
and \ref{hilum_color_all_T_kmers}.

Although \cite{bandillo2017} listed the \textit{W1} locus (typically associated
with flower color) as putatively associated with hilum color, we found no
evidence of association in our results.

The second GWAS analysis contrasting accessions with either black or brown hilum
color was better suited at resolving the signal at the \emph{R} locus (Table
\ref{HILUM.COLOR_gwas_table}). Accordingly, this GWAS analysis detected strong
signals overlapping the \emph{R} locus and no other classical loci using all
four approaches (Figure \ref{hilum_color_blackbrown_manhattan}).
\cite{gillman2011} identified the R2R3 MYB transcription factor Glyma.09g235100
as the molecular basis for the \emph{R} locus and documented four different
loss-of-function alleles linked to brown hilum color. Although our analyses did
detect strong signals in the vicinity of this gene, the most associated markers
were not located in the body of the gene (Figure
\ref{hilum_color_blackbrown_R_signal}). Interestingly, the most significantly
associated \emph{k}-mers detected by our analysis mapped 2 kb upstream of
another gene (Glyma.09g234900) annotated as a MYB transcription factor and
putatively involved in the regulation of anthocyanin biosynthesis. This gene
as well as Glyma.09g235100 and two others are part of a tandem array of four
R2R3 MYB genes considered as candidates by \cite{gillman2011}, and of which
Glyma.09g235100 was the only one expressed in the seed coat. Given the
convincing evidence provided by \cite{gillman2011}, it is most likely that
Glyma.09g235100 is indeed the gene associated with the \emph{R} locus and that
our failure to identify the most significantly associated \emph{k}-mers as
causal variants is due to the coexistence of several variants that can cause
loss of function at the \emph{R} locus. Indeed, we observed at least two causal
variants (a 1-bp deletion and a SNP) previously documented by
\cite{gillman2011} in our dataset (Figure \ref{hilum_color_blackbrown_R_gene},
\ref{hilum_color_blackbrown_R_kmers}).

We conducted a third GWAS focusing on the brown and reddish brown hilum color
phenotypes in order to detect a signal at the \emph{O} locus, however this GWAS
did not detect associations at any of the classical loci (Figure
\ref{hilum_color_rbr_manhattan}).

\subsection*{Pod color}
\label{sv-gwas-pod-color}

We conducted two different GWAS analyses on pod color. The first analysis included four
different colors (Table \ref{POD.COLOR_gwas_table}) and resulted in the
detection of signals overlapping both the \emph{L1} and \emph{L2} loci using all
four methods (Figure \ref{pod_color_all_manhattan}). Detailed results at the
\textit{L1} locus can be viewed in figure \ref{pod_color_all_L1_signal},
whereas detailed results at the \textit{L2} locus can be viewed in figure
\ref{pod_color_all_L2_signal}.

The second GWAS analysis compared black and brown pods with the objective of detecting stronger
signals at the \emph{L1} locus (Table
\ref{POD.COLOR_gwas_table}).  Accordingly, all four methods detected a signal
at the \emph{L1} locus but not at the \emph{L2} locus using this GWAS (Figure
\ref{pod_color_blbr_manhattan}).  Results at the \emph{L1} locus were similar
to those observed when using all observed phenotypes, but the \emph{k}-mer
analysis revealed a previously undocumented signal on chromosome Gm15 which was
not in LD with any other signal (Figures \ref{pod_color_blbr_manhattan},
\ref{pod_color_blbr_ld}). The observed signal at this locus overlaps a total of
27 genes (Figure \ref{pod_color_blbr_pdcGm15_signal}), but none of those are
annotated as having an obvious role in controlling pod color.

\subsection*{Pubescence form}
\label{sv-gwas-pubesence-form}

We conducted two different GWAS analyses on pubescence form. The first analysis used three
phenotypes (erect, semi-appressed and appressed; Table
\ref{PUBESCENCE.FORM_gwas_table}) and detected signals overlapping the
\emph{Pa1} locus using all four approaches as well as signals overlapping the
\emph{Pa2} locus using Platypus and the \emph{k}-mer approach (Figure
\ref{pubescence_form_all_manhattan}). Detailed results at the \textit{Pa1} locus
can be viewed in figure \ref{pubescence_form_all_Pa1_signal},
\ref{pubescence_form_all_Pa1_gene} and \ref{pubescence_form_all_Pa1_kmers},
whereas results at the \textit{Pa2} locus can be viewed in figure
\ref{pubescence_form_all_Pa2_signal}.

Signals were also observed on chromosomes Gm04 and Gm15 by the Platypus and
\emph{k}-mer analyses (Figure \ref{pubescence_form_all_manhattan}), however
upon further analysis of LD patterns they appeared to be linked to the
\emph{Pa1} locus (Figure \ref{pubescence_form_all_ld}).

The second GWAS analysis focused only on appressed and semi-appressed phenotypes
with the objective of enhancing the signal at the \emph{Pa2} locus. However,
this analysis only detected weak signals at the \emph{Pa1} locus (Figure
\ref{pubescence_form_noerect_manhattan}) and its results were therefore not
analyzed further.

\subsection*{Pubescence density}
\label{sv-gwas-pubescence-density}

We conducted a GWAS analysis on pubescence density (Table
\ref{PUBESCENCE.DENSITY_gwas_table}), which detected signals at the \emph{Ps}
locus but not the \emph{Pd1} and \emph{P1} loci using all four approaches
(Figure \ref{pubescence_density_manhattan}). \cite{liu2020ps} linked the gene
Glyma.12g187200 to the \emph{Ps} locus and identified ia 25.6-kb copy number variation
containing this gene as the causal variant, with
higher copy number resulting in decreased pubescence. By
analyzing the sequencing depth in accessions with increased copy number, we
identified the boundaries of the CNV region as 36,254,989 -
36,297,368 (42 kb) on chromosome Gm12.
Detailed results at this locus can be viewed in figures \ref{pubescence-density-main-figure},
\ref{pubescence_density_Ps_signal} and \ref{pubescence_density_Ps_gene}.

\subsection*{Seed coat luster}
\label{sv-gwas-seed-coat-luster}

We performed three different GWAS analyses on seed coat luster (Table
\ref{SEED.COAT.LUSTER_gwas_table}). Since all three analyses yielded similar results
(Figure \ref{seed_coat_luster_all_manhattan},
\ref{seed_coat_luster_nointermediate_manhattan},
\ref{seed_coat_luster_dullshiny_manhattan}), we chose to focus on the
GWAS contrasting the dull and shiny phenotypes as it yielded the most
significant signals. In addition, it also provided results that were most consistent with the
analysis previously done by \cite{bandillo2017}. The analysis detected
signals at the \emph{B} locus as the sole classical locus using all four
methods (Figure \ref{seed_coat_luster_dullshiny_manhattan}). Apparent signals
on chromosomes Gm20 and Gm09 (which may correspond to the B? locus reported by
\cite{bandillo2017}) were also detected, but analysis of the LD patterns
between \emph{k}-mers suggested that these signals were most likely in LD with
the \emph{B} locus (Figure \ref{seed_coat_luster_dullshiny_ld}). All methods
also detected signals overlapping the \emph{Hps} locus as defined by
\cite{bandillo2017}, however results shown by \cite{gijzen2003} and our results
suggest that \emph{Hps} and \emph{B} are the same locus. Using a sequencing depth-based segmentation analysis
similar to the one used for the \emph{Ps} locus, we identified a 31-kb region
spanning positions 9,386,109-9,417,431 on chromosome Gm15 that exhibited
variation in copy number across accessions and overlapped four genes annotated
as containing HPS domains (Figures \ref{seed_coat_luster_dullshiny_B_signal},
\ref{seed_coat_luster_dullshiny_B_gene}). The top 5\% / 1\% significant
associations of all approaches other than vg overlapped this CNV region and the
most significantly associated \emph{k}-mer was located within the CNV region
(Figures \ref{seed_coat_luster_dullshiny_B_signal},
\ref{seed_coat_luster_dullshiny_B_gene}). However we were not able to link any
of the variants found to the presence of copy number variation.

% Seed coat luster - I locus

\cite{bandillo2017} mentioned the \textit{I} locus as possibly associated
with seed coat luster, however we did not find evidence for such an association
in our dataset.

\subsection*{Maturity group}
\label{sv-gwas-maturity-group}

We conducted on a GWAS analysis on maturity group which did not yield any significant signals
overlapping the classical E1-E4 loci using any of the four GWAS approaches.
However, signals at the E1 locus did approach significance in the Platypus
analysis (Table \ref{MATURITY.GROUP_gwas_table}, Figure \ref{maturity_group_manhattan}).

\subsection*{Resistance to \emph{Phytophthora sojae}}
\label{sv-gwas-resistance}

We conducted a GWAS analysis on horizontal resistance to
\textit{Phytophthora sojae} using corrected dry weight (CDW) as a quantitative
measure of resistance. Detailed results can be viewed in figures
\ref{corrected_dry_weight_manhattan} and
\ref{corrected_dry_weight_cdwGm15_signal}.

\subsection*{Seed oil content}
\label{sv-gwas-seed-oil-content}

We conducted a GWAS analysis on seed oil content using data obtained from the
GRIN database. While several seed oil content quantitative trait loci (QTL)
have been reported due to high interest in this trait for soybean breeding
\citep{chaudhary2015}, we focused on two major QTL on chromosomes Gm15 and Gm20.
These loci have often been reported in \emph{G. max} and were notably observed by
\cite{bandillo2015} on a collection of USDA lines using SoySNP50K chip
genotypes and GRIN phenotypic data. We identified a signal located on
chromosome Gm15 using the \emph{k}-mer approach only (Figure
\ref{oil_manhattan}). Although this signal did not overlap with the coordinates
found by \cite{bandillo2017}, it was located only 2 kb away from it and 30 kb
away from Glyma.15g049200, the sugar transporter gene identified by
\cite{zhang2020} as associated with that locus. Therefore, the significantly
associated \emph{k}-mers were not linked to the causal variation at that locus.

\subsection*{Seed protein content}
\label{sv-gwas-seed-protein-content}

Similarly to seed oil content, we conducted a GWAS on seed protein content
using data obtained from the GRIN database. As is generally the case in
soybean given the negative correlation between seed oil and protein content,
the two major QTL considered above and reported by \cite{bandillo2015} for seed
oil content are also major QTL for seed protein content. We only identified a signal
at the locus located on chromosome Gm20, and only using the genotypes
called by vg (Figure \ref{protein_manhattan}). However, the SV discovered as
significantly associated was located about 700 kb from the gene cloned by
\cite{fliege2022} for that locus (Glyma.20g085100). It is therefore unlikely that this SV
represents the causal variation at that locus.

\clearpage

% Supplemental tables
\section*{Supplemental tables}
\label{annexe-sv-gwas-tables}

\clearpage

\gwastableone{FLOWER.COLOR}{flower color}
\gwastabletwo{PUBESCENCE.COLOR}{pubescence color}
\gwastabletwo{SEED.COAT.COLOR}{seed coat color}
\gwastabletwo{STEM.TERMINATION.TYPE}{stem termination type}
\gwastablethree{HILUM.COLOR}{hilum color}
\gwastabletwo{POD.COLOR}{pod color}
\gwastabletwo{PUBESCENCE.FORM}{pubescence form}
\gwastableone{PUBESCENCE.DENSITY}{pubescence density}
\gwastablethree{SEED.COAT.LUSTER}{seed coat luster}
\gwastableone{MATURITY.GROUP}{maturity group}


% Table with all signals
\begin{landscape} \footnotesize
	\begin{ThreePartTable}

		\begin{TableNotes}
		\item[a] Most significant -log\textsubscript{10}(\emph{p}) previously reported at this locus by the study in the Study column
		\item[b] Gene associated with the locus. A question mark following the identifier of the gene indicates that this is simply a candidate
			that has yet to be confirmed. A dash indicates that no candidate has been suggested yet. We have only included candidate
			genes that have been suggested in previous studies.
		\item[c] Most significant -log\textsubscript{10}(\emph{p}) reported by Platypus at this locus. A dash indicates that no signal
			was detected by Platypus at this locus. The same applies for other methods (vg, Paragraph, \emph{k}-mers) in
			their respective columns.
		\item[d] The \emph{p}-value at the W1 locus using the \emph{k}-mers method was so small that it was numerically equal
			to zero. This value was therefore set to the -log\textsubscript{10} of the smallest value that could be represented
			in R using our machine.
		\end{TableNotes}

		\begin{longtable}{llcclccccl}

			\caption[Description of loci associated with the traits studied]%
			{Description of loci associated with the traits studied and -log\textsubscript{10}(\emph{p}) of significant
			signals detected in our study using four different genotypic datasets (Platypus, vg, Paragraph, \emph{k}-mers).}\\
			\label{signals_table}\\

			\toprule
			Trait & Locus & Chromosome & -log\textsubscript{10}(\emph{p})\tnote{a} & Gene\tnote{b} & Platypus\tnote{c} & vg & Paragraph & \emph{k}-mers  & Study \\
			\midrule \endfirsthead

			\caption[]{Loci associated with the traits studied \emph{(continued)}}\\

			\toprule
			Trait & Locus & Chromosome & -log10(p) & Gene & Platypus & vg & Paragraph & k-mers  & Study \\
			\midrule
			\endhead

			\bottomrule
			Continued on next page \\
			\endfoot

			\bottomrule \insertTableNotes
			\endlastfoot
			\csvreader[head to column names,
			late after line=\\,
			late after last line=,
			before reading={\catcode`\#=12},
			after reading={\catcode`\#=6},
			/csv/separator=semicolon]%
			{../tables/signals_table.csv}{}{\Trait & \Locus & \Chromosome & \Pvalues & \Gene & \Platypus & \Vg & \Paragraph & \Kmers & \Study}

		\end{longtable}

	\end{ThreePartTable}

\end{landscape}
\clearpage

% Supplemental figures
\section*{Supplemental figures}
\label{annexe-sv-gwas-figures}

% Figures mentioned in the main text
% Figures for flower color
\manhattanplot{flower_color}{flower color}{\getvar{flower_color_nsamples}}
\signalplot{flower_color_W1}{flower color}{\emph{W1}}%
{Vertical dotted lines indicate the location of the \getvar{flower_color_W1_gene} gene associated with the locus.}
\geneplot{flower_color_W1}{flower color}{\emph{W1}}{\getvar{flower_color_W1_gene} gene}%
{Vertical dotted lines in panels (c), (d) and (e) indicate the location of the causal SV at this locus.}
%\kmerplot{flower_color_W1}

% Figures for pubescence color
\manhattanplot{pubescence_color_all}{pubescence color (first GWAS)}{\getvar{pubescence_color_all_nsamples}}
\signalplot{pubescence_color_all_T}{pubescence color (first GWAS)}{\emph{T}}%
{Vertical dotted lines indicate the location of the \getvar{pubescence_color_all_T_gene} gene associated with the locus.}
\geneplot{pubescence_color_all_T}{pubescence color (first GWAS)}{\emph{T}}{\getvar{pubescence_color_all_T_gene} gene}%
{Vertical dotted lines in panels (b) and (e) indicate the location of the causal indel at this locus.}
\kmerplot{pubescence_color_all_T}{\getvar{pubescence_color_all_T_gene}}{\emph{T}}{pubescence color (first GWAS)}{a causal indel}

\manhattanplot{pubescence_color_nogray}{pubescence color (second GWAS)}{\getvar{pubescence_color_nogray_nsamples}}
\signalplot{pubescence_color_nogray_Td}{pubescence color (second GWAS)}{\emph{Td}}%
{Vertical dotted lines indicate the location of the \getvar{pubescence_color_nogray_Td_gene} gene associated with the locus.}
\geneplot{pubescence_color_nogray_Td}{pubescence color (second GWAS)}{\emph{Td}}{\getvar{pubescence_color_nogray_Td_gene} gene}%
{Vertical dotted lines in panels (b) and (e) indicate the location of the causal SNP at this locus.}
%\kmerplot{pubescence_color_nogray_Td}

% Figures for seed coat color
\manhattanplot{seed_coat_color_all}{seed coat color (first GWAS)}{\getvar{seed_coat_color_all_nsamples}}
\signalplot{seed_coat_color_all_I}{seed coat color (first GWAS)}{\emph{I}}%
{Vertical dotted lines indicate the boundaries of the tandem duplication/inversion identified as the causal variant at this locus.}

\manhattanplot{seed_coat_color_greenyellow}{seed coat color (second GWAS)}{\getvar{seed_coat_color_greenyellow_nsamples}}
\signalplot{seed_coat_color_greenyellow_G}{seed coat color (second GWAS)}{\emph{G}}%
{Vertical dotted lines indicate the location of the \getvar{seed_coat_color_greenyellow_G_gene} gene associated with the locus.}
\geneplot{seed_coat_color_greenyellow_G}{seed coat color (second GWAS)}{\emph{G}}{\getvar{seed_coat_color_greenyellow_G_gene} gene}%
{Vertical dotted lines in panels (b) and (e) indicate the location of the causal SNP at this locus.}
%\kmerplot{seed_coat_color_greenyellow_G}

% Figures for stem termination
\manhattanplot{stem_termination_all}{stem termination type (first GWAS)}{\getvar{stem_termination_all_nsamples}}
\signalplot{stem_termination_all_Dt1}{stem termination type (first GWAS)}{\emph{Dt1}}%
{Vertical dotted lines indicate the location of the \getvar{stem_termination_all_Dt1_gene} gene associated with the locus.}
\geneplot{stem_termination_all_Dt1}{stem termination type (first GWAS)}{\emph{Dt1}}{\getvar{stem_termination_all_Dt1_gene} gene}%
{Vertical dotted lines in panels (b) and (e) indicate the location of a causal SNP at this locus.}
\kmerplot{stem_termination_all_Dt1}{\getvar{stem_termination_all_Dt1_gene}}{\emph{Dt1}}{stem termination type (first GWAS)}{a causal SNP}


% Figures for hilum color
\manhattanplot{hilum_color_all}{hilum color (first GWAS)}{\getvar{hilum_color_all_nsamples}}
\signalplot{hilum_color_all_T}{hilum color (first GWAS)}{\emph{T}}%
{Vertical dotted lines indicate the location of the \getvar{hilum_color_all_T_gene} gene associated with the locus.}
\geneplot{hilum_color_all_T}{hilum color (first GWAS)}{\emph{T}}{\getvar{hilum_color_all_T_gene} gene}%
{Vertical dotted lines in panels (b) and (e) indicate the location of the causal indel at this locus.}
\kmerplot{hilum_color_all_T}{\getvar{hilum_color_all_T_gene}}{\emph{T}}{hilum color (first GWAS)}{a causal indel}
\signalplot{hilum_color_all_I}{hilum color (first GWAS)}{\emph{I}}%
{Vertical dotted lines indicate the boundaries of the tandem duplication/inversion identified as the causal variant at this locus.}

\manhattanplot{hilum_color_blackbrown}{hilum color (second GWAS)}{\getvar{hilum_color_blackbrown_nsamples}}
\signalplot{hilum_color_blackbrown_R}{hilum color (second GWAS)}{\emph{R}}%
{Vertical dotted lines indicate the location of the \getvar{hilum_color_blackbrown_R_gene} gene associated with the locus.}
\geneplot{hilum_color_blackbrown_R}{hilum color (second GWAS)}{\emph{R}}{\getvar{hilum_color_blackbrown_R_gene} gene}%
{Vertical dotted lines indicate the location an indel (panels b and e) and a SNP (panel e) that are putatively causal at this locus.}
\kmerplot{hilum_color_blackbrown_R}{\getvar{hilum_color_blackbrown_R_gene}}{\emph{R}}{hilum color (second GWAS)}{a SNP and an indel that are putatively causal}

% Figures for pubescence density
\manhattanplot{pubescence_density}{pubescence density}{\getvar{pubescence_density_nsamples}}

% Figures for seed coat luster
\manhattanplot{seed_coat_luster_dullshiny}{seed coat luster (third GWAS)}{\getvar{seed_coat_luster_dullshiny_nsamples}}
\signalplot{seed_coat_luster_dullshiny_B}{seed coat luster (third GWAS)}{\emph{B}}%
{Vertical dotted lines indicate the boundaries of the causal CNV associated with the locus.}

% Figures for maturity group
\manhattanplot{maturity_group}{maturity group}{\getvar{maturity_group_nsamples}}

% Figures for oil
\manhattanplot{oil}{seed oil content}{\getvar{oil_nsamples}}

% Figures for protein
\manhattanplot{protein}{seed protein content}{\getvar{protein_nsamples}}

% Figures for pod color
\manhattanplot{pod_color_all}{pod color (first GWAS)}{\getvar{pod_color_all_nsamples}}
\signalplot{pod_color_all_L1}{pod color (first GWAS)}{\emph{L1}}{}
\signalplot{pod_color_all_L2}{pod color (first GWAS)}{\emph{L2}}{}

% Figures for pubescence form
\manhattanplot{pubescence_form_all}{pubescence form (first GWAS)}{\getvar{pubescence_form_all_nsamples}}
\signalplot{pubescence_form_all_Pa1}{pubescence form (first GWAS)}{\emph{Pa1}}%
{Vertical dotted lines indicate the location of the \getvar{pubescence_form_all_Pa1_gene} gene that we suggest as a candidate for this locus.}
\geneplot{pubescence_form_all_Pa1}{pubescence form (first GWAS)}{\emph{Pa1}}{\getvar{pubescence_form_all_Pa1_gene} candidate gene putatively}%
{Vertical dotted lines in panels (b) and (e) indicate the location of two highly significant non-synonymous SNPs at this gene.}
\kmerplot{pubescence_form_all_Pa1}{\getvar{pubescence_form_all_Pa1_gene} candidate}{\emph{Pa1}}{pubescence form (first GWAS)}{a non-synonymous SNP}
\signalplot{pubescence_form_all_Pa2}{pubescence form (first GWAS)}{\emph{Pa2}}{}

% Figures for corrected dry weight
\manhattanplot{corrected_dry_weight}{corrected dry weight (resistance to \emph{P. sojae})}{\getvar{corrected_dry_weight_nsamples}}
\signalplot{corrected_dry_weight_cdwGm15}{corrected dry weight (resistance to \emph{P. sojae})}{\emph{cdwGm15}}%
{Vertical dotted lines indicate the location of the \getvar{corrected_dry_weight_cdwGm15_gene} gene suggested by de Ronne et al. (2022) as associated with the locus.}
%\ldplot{corrected_dry_weight}

% Figures for novel loci
\manhattanplot{pod_color_blbr}{pod color (second GWAS)}{\getvar{pod_color_blbr_nsamples}}
\ldplot{pod_color_blbr}{pod color (second GWAS)}{\getvar{pod_color_blbr_ldkmers}}

% Figures mentioned only in supplemental data

% Histogram with concordance rates
\begin{cfigure}
	\includegraphics[width = 6in]{concordance_histogram}
	\caption{Histogram of concordance rates between genotypes derived from WGS data and SoySNP50K genotypes for 385 soybean accessions.
	The vertical line indicates the 0.9 concordance threshold that was used for filtering out mismatching samples.}
	\label{concordance-histogram}

\end{cfigure}

\clearpage

% Flower color
\ldplot{flower_color}{flower color}{\getvar{flower_color_ldkmers}}

% Pubescence color
\ldplot{pubescence_color_all}{pubescence color (first GWAS)}{\getvar{pubescence_color_all_ldkmers}}

% Stem termination
\manhattanplot{stem_termination_sn}{stem termination type (second GWAS)}{\getvar{stem_termination_sn_nsamples}}
%\ldplot{stem_termination_sn}

% Hilum color
\manhattanplot{hilum_color_rbr}{hilum color (third GWAS)}{\getvar{hilum_color_rbr_nsamples}}

% Pod color
\signalplot{pod_color_blbr_pdcGm15}{pod color (second GWAS)}{newly suggested \emph{pdcGm15}}{}

% Pubescence form
\ldplot{pubescence_form_all}{pubescence form (first GWAS)}{\getvar{pubescence_form_all_ldkmers}}
\manhattanplot{pubescence_form_noerect}{pubescence form (second GWAS)}{\getvar{pubescence_form_noerect_nsamples}}

% Pubescence density
\signalplot{pubescence_density_Ps}{pubescence density}{\emph{Ps}}%
{Vertical dotted lines indicate the boundaries of the causal CNV overlapping the \getvar{pubescence_density_Ps_gene} gene associated with the locus.}
\geneplot{pubescence_density_Ps}{pubescence density}{\emph{Ps}}{location of the causal CNV overlapping the \getvar{pubescence_density_Ps_gene} gene.}%
{The vertical dotted line in panel (e) indicates the location of the most significant \emph{k}-mer, which is associated with the causal CNV at this locus.
The shaded gray rectangles show the boundaries of the causal CNV.}

% Seed coat luster
\manhattanplot{seed_coat_luster_all}{seed coat luster (first GWAS)}{\getvar{seed_coat_luster_all_nsamples}}
\manhattanplot{seed_coat_luster_nointermediate}{seed coat luster (second GWAS)}{\getvar{seed_coat_luster_nointermediate_nsamples}}

\ldplot{seed_coat_luster_dullshiny}{seed coat luster (third GWAS)}{\getvar{seed_coat_luster_dullshiny_ldkmers}}

\geneplot{seed_coat_luster_dullshiny_B}{seed coat luster (third GWAS)}{\emph{B}}{location of the causal CNV}%
{The shaded gray rectangles show the boundaries of the causal CNV.}

% Bibliography
\bibliographystyle{genome_research}              % style de la bibliographie
\bibliography{references}                   % production de la bibliographie

