\documentclass[12pt]{article}

\usepackage[utf8]{inputenc} % declare encoding as utf8
\usepackage{graphicx} % to enable \includegraphics
\usepackage{flafter} % to make sure floats appear after their position in text
\usepackage{pdflscape} % allow some pages to be in landscape orientation
\usepackage{layout} % for debugging purposes, looking at page layout
\usepackage[margin=1in]{geometry} % leaves more space for text body
\usepackage[hidelinks]{hyperref} % hyperlinks from list of figures/tables
\usepackage[labelfont=bf]{caption} % format captions and make links go to figure instead of caption
\usepackage{csvsimple} % read and display csv files as tables
\usepackage{booktabs} % prettier tables
\usepackage{threeparttable} % easy addition of footnotes below tables
\usepackage{textcomp} % makes the \textdegree command available
\usepackage{tocloft} % formatting the lists of tables and figures
\usepackage{xcolor} % enable text coloring
\usepackage{calc} % makes \widthof available to set the width of the last column
\usepackage{array} % using command \arraybackslash to restore the definition of \\ after \centering
\usepackage{longtable} % allow tables to span more than one page
\usepackage{threeparttablex} % threeparttable for longtable
\usepackage{capt-of} % put caption on next page
\usepackage{listings}
\usepackage{datatool} % allows fetching key-value pairs from a text file

% Setting the separator to use with datatool
\DTLsetseparator{ = }

% Formatting the table of contents/subsection
\renewcommand{\contentsname}{Supplemental methods}

% Formatting the list of tables
\renewcommand{\listtablename}{Supplemental tables} % Write "Supplemental tables" instead of "List of tables"
\renewcommand\cfttabpresnum{Table } % Write "Table" before table number in list of tables
\setlength{\cfttabnumwidth}{6em} % Spacing before captions
\setlength{\cftbeforefigskip}{1ex} % Vertical spacing between entries

% Formatting the list of figures
\renewcommand{\listfigurename}{Supplemental figures} % Write "Supplemental figures" instead of "List of figures"
\renewcommand\cftfigpresnum{Figure } % Write "Figure" before fig number in list of figures
\setlength{\cftfignumwidth}{6em} % Spacing before captions
\setlength{\cftbeforetabskip}{1ex} % Vertical spacing between entries

\graphicspath{ {../figures/} } % tell Latex to look for figures in figures/

% Format figure and table numbers to follow S1, S2, ...
\renewcommand\thefigure{S\arabic{figure}} 
\renewcommand\thetable{S\arabic{table}} 

% Defining new environments as I will always want the figures centered
\newenvironment{cfigure}
	{\begin{figure} \centering}
	{\end{figure}}

\newenvironment{lsfigure}
	{\begin{landscape} \begin{figure} \centering}
	{\end{figure} \end{landscape}}

\setcounter{totalnumber}{1} % Allow a maximum of one float (table/figure) per page

\setlength{\parskip}{2ex}

% Defining a command to generate a table of phenotypic data for a single gwas
\newcommand{\gwastableone}[2]{
\begin{table}
	\caption[Phenotypic data used for the GWAS analysis of #2]%
	{Phenotype frequency and numerical coding of phenotypic data used for the GWAS analysis of #2.
	``NA'' indicates a missing value in the source data.
	Dashes indicate that this phenotype was not used in the GWAS analysis.}
	\label{#1_gwas_table}

	\centering
	\begin{threeparttable}\small

		\csvreader[head to column names,
		/csv/separator=semicolon,
		tabular = llc,
		table head = \toprule Phenotype & Frequency & GWAS \\ \midrule,
		table foot = \bottomrule]%
		{../tables/#1_gwas_table.csv}%
		{}%
		{\Value & \Frequency & \GWASa}

	\end{threeparttable}

\end{table}
\clearpage
}

% Defining a command to generate a table of phenotypic data for two gwas
\newcommand{\gwastabletwo}[2]{
\begin{table}
	\caption[Phenotypic data used for the GWAS analyses of #2]%
	{Phenotype frequency and numerical coding of phenotypic data used in two GWAS analyses of #2.
	``NA'' indicates a missing value in the source data.
	Dashes indicate that this phenotype was not used in the GWAS analysis.}
	\label{#1_gwas_table}

	\centering
	\begin{threeparttable}\small

		\csvreader[head to column names,
		/csv/separator=semicolon,
		tabular = llcc,
		table head = \toprule Phenotype & Frequency & GWAS \#1 & GWAS \# 2\\ \midrule,
		table foot = \bottomrule]%
		{../tables/#1_gwas_table.csv}%
		{}%
		{\Value & \Frequency & \GWASa & \GWASb}

	\end{threeparttable}

\end{table}
\clearpage
}

% Defining a command to generate a table of phenotypic data for three gwas
\newcommand{\gwastablethree}[2]{
\begin{table}
	\caption[Phenotypic data used for the GWAS analyses of #2]%
	{Phenotype frequency and numerical coding of phenotypic data used in three GWAS analyses of #2.
	``NA'' indicates a missing value in the source data.
	Dashes indicate that this phenotype was not used in the GWAS analysis.}
	\label{#1_gwas_table}

	\centering
	\begin{threeparttable}\small

		\csvreader[head to column names,
		/csv/separator=semicolon,
		tabular = llccc,
		table head = \toprule Phenotype & Frequency & GWAS \#1 & GWAS \# 2 & GWAS \# 3\\ \midrule,
		table foot = \bottomrule]%
		{../tables/#1_gwas_table.csv}%
		{}%
		{\Value & \Frequency & \GWASa & \GWASb & \GWASc}

			   \begin{tablenotes}
			   \item[a] test
			   \end{tablenotes}
	\end{threeparttable}

\end{table}
\clearpage
}

% Defining a command for manhattan plot figures
\newcommand{\manhattanplot}[3]{
\begin{cfigure}
	\includegraphics[width = 6in]{#1_manhattan}
	\caption[Manhattan plot of #2]{Manhattan plots generated from the GWAS analysis of #2 on #3 samples using four genotype datasets :
	(a) Platypus (SNPs and indels), (b) vg (SVs), (c) Paragraph (SVs), (d) \emph{k}-mers presence/absence.
	The x-axis shows the position along the reference assembly version 4 of Williams82.
	Each point shows the $-\log_{10} {p-value}$ associated with a particular marker or \emph{k}-mer.
	Horizontal dotted lines indicate the 5\% family-wise error-rate threshold determined from a randomization approach.
	Vertical dotted lines indicate the position of signals associated with the trait.
	Signals documented in previous studies are colored according to whether they were founs (green) or not (red) by a particular approach.
	The ``Gm'' prefix has been left out of chromosome names for simplicity.
	}
	\label{#1_manhattan}
\end{cfigure}

\clearpage}

% Defining a new command for signal plot figures
\newcommand{\signalplot}[4]{
\begin{cfigure}
	\includegraphics[width = 6in]{#1_signal}
	\caption[Zoomed-in Manhattan plot at #3 locus for #2]{Zommed-in Manhattan plots generated from the GWAS analysis of #2 at the #3 locus using four genotype datasets:
	(b) Platypus (SNPs and indels), (c) vg (SVs), (d) Paragraph (SVs), (e) \emph{k}-mers presence/absence.
	Panel (a) shows the location of annotated gene models over the plotting interval.
	#4
	Gray shaded rectangles indicate the region delimited by the top 5\% (Paragraph and vg) or top 1\% (Platypus and \emph{k}-mers) associations in the signal region.
	}
	\label{#1_signal}
\end{cfigure}

\clearpage}

% Defining a new command for gene plot figures
\newcommand{\geneplot}[4]{
\begin{cfigure}
	\includegraphics[width = 6in]{#1_gene}
	\caption[Zommed-in Manhattan plot at gene(s) associated with #3 locus for #2]%
	{Zoomed-in Manhattan plots generated from the GWAS analysis of #2 at the #4 associated with the #3 locus using four genotype datasets:
	(b) Platypus (SNPs and indels), (c) vg (SVs), (d) Paragraph (SVs), (e) \emph{k}-mers presence/absence.
	Panel (a) shows the location of annotated gene models over the plotting interval.
	Exons are represented by rectangles whereas introns are represented by horizontal lines.
	Coding sequences are shown in blue and the direction of translation is indicated by arrows.
	}
	\label{#1_gene}
\end{cfigure}

\clearpage}

% Defining a new command for k-mer plot figures
\newcommand{\kmerplot}[4]{
	\begin{cfigure}
		\includegraphics[width = 6.5in]{#1_kmers}
		\caption[Variants associated with significant \emph{k}-mers at the #3 locus]%
		{Identification of variants underlying significant \emph{k}-mers at the #2 associated with the #3 locus for #4.
		(a) Gene model of #2. Exons are represented by rectangles whereas introns are represented by horizontal lines.
		Coding sequences are shown in blue and the direction of translation is indicated by arrows.
		The red rectangle identifies the region that is zoomed-in in panel (b).
		(b) Nucleotide sequences of haplotypes observed in at least five samples across the dataset.
		Individual nucleotides are colored according to the $-\log_{10}(p-value)$ of the most significant \emph{k}-mer overlapping them.
		Dashes indicate gaps in haplotype sequence alignment whereas vertical lines indicate differences in sequence between two haplotypes.
		(c) Contingency table of the phenotypes and haplotypes observed in the dataset. Haplotypes corresponds to those shown in panel (b).}
		\label{#1_kmers}
	\end{cfigure}

\clearpage}

% Defining a new command for LD plot figures
\newcommand{\ldplot}[3]{
	\begin{cfigure}
		\includegraphics[width = 6.5in]{#1_ld}
		\caption[LD plot of significant \emph{k}-mers for #2]{Pairwise LD among #3 significant \emph{k}-mers identified for #2.
		\emph{k}-mers are sorted along the y-axis according to their putative position along the reference assembly version 4 of Williams82,
		as identified by ``Gm'' chromosome labels.
		Sequences that lack a ``Gm'' label (top of the y-axis) represent unanchored scaffolds.
		\emph{k}-mers are represented in the same order along the x- and y-axis.
		The colored rectangles drawn below the x-axis represent the $-\log_{10} (p-values)$ of each \emph{k}-mer.}
		\label{#1_ld}
	\end{cfigure}

\clearpage}

% Allows underscores in captions but is not a great solution
\catcode`_=12

% Information for the title page
\title{Supplemental data to ``ARTICLE TITLE''}
\author{Marc-André Lemay \and François Belzile}
\date{}

% END OF PREAMBLE

\begin{document}

% Reading the contents of the variables.txt file first
\DTLloaddb[noheader, keys={key,value}]{variables}{variables.txt}
% Defining a new command to fetch the value from the key
\newcommand{\getvar}[1]{\DTLfetch{variables}{key}{#1}{value}}

\maketitle \thispagestyle{empty}

\tableofcontents

\vspace{4ex}

\listoftables

\thispagestyle{empty}
%\clearpage

\vspace{4ex}

\listoffigures

%\thispagestyle{empty}

\clearpage

\begin{landscape} \footnotesize
	\begin{ThreePartTable}

		\begin{TableNotes}
		\item[a] Most significant -log\textsubscript{10}(p) reported at this locus by the study in the Study column
		\item[b] Gene associated with the locus. An question mark following the identifier of the gene indicates that this is simply a candidate
			which has yet to be confirmed. A dash indicates that no candidate has been suggested yet.
		\item[c] Most significant -log\textsubscript{10}(p) reported by Platypus at this locus. A dash indicates that no signal
			was detected by Platypus at this locus. The same applies for other methods (vg, Paragraph, \emph{k}-mers) in
			their respective columns.
		\end{TableNotes}

		\begin{longtable}{llcclccccl}

			\caption[Description of loci associated with the traits studied]%
			{Description of loci associated with the traits studied and -log\textsubscript{10}(p) of significant
			signals detected in our study using four different genotypic datasets.}\\
			\label{signals_table}\\

			\toprule
			Trait & Locus & Chromosome & -log\textsubscript{10}(p)\tnote{a} & Gene\tnote{b} & Platypus\tnote{c} & vg & Paragraph & \emph{k}-mers  & Study \\
			\midrule \endfirsthead

			\caption[]{Loci associated with the traits studied \emph{(continued)}}\\

			\toprule
			Trait & Locus & Chromosome & -log10(p) & Gene & Platypus & vg & Paragraph & k-mers  & Study \\
			\midrule
			\endhead

			\bottomrule
			Continued on next page \\
			\endfoot

			\bottomrule \insertTableNotes
			\endlastfoot
			\csvreader[head to column names,
			late after line=\\,
			late after last line=,
			before reading={\catcode`\#=12},
			after reading={\catcode`\#=6},
			/csv/separator=semicolon]%
			{../tables/signals_table.csv}{}{\Trait & \Locus & \Chromosome & \Pvalues & \Gene & \Platypus & \Vg & \Paragraph & \Kmers & \Study}

		\end{longtable}

	\end{ThreePartTable}

\end{landscape}
\clearpage

\gwastableone{FLOWER.COLOR}{flower color}
\gwastabletwo{PUBESCENCE.COLOR}{pubescence color}
\gwastabletwo{SEED.COAT.COLOR}{seed coat color}
\gwastabletwo{STEM.TERMINATION.TYPE}{stem termination type}
\gwastablethree{HILUM.COLOR}{hilum color}
\gwastabletwo{POD.COLOR}{pod color}
\gwastabletwo{PUBESCENCE.FORM}{pubescence form}
\gwastableone{PUBESCENCE.DENSITY}{pubescence density}
\gwastablethree{SEED.COAT.LUSTER}{seed coat luster}
\gwastableone{MATURITY.GROUP}{Maturity group}

% Figures for flower color
\manhattanplot{flower_color}{flower color}{\getvar{flower_color_nsamples}}
\ldplot{flower_color}{flower color}{\getvar{flower_color_ldkmers}}
\signalplot{flower_color_W1}{flower color}{\emph{W1}}%
{Vertical dotted lines indicate the location of the \getvar{flower_color_W1_gene} gene associated with the locus.}

\geneplot{flower_color_W1}{flower color}{\emph{W1}}{\getvar{flower_color_W1_gene} gene}
\kmerplot{flower_color_W1}{\getvar{flower_color_W1_gene} gene}{\emph{W1}}{flower color}


% Figures for pubescence color
\manhattanplot{pubescence_color_all}{pubescence color (first GWAS)}{\getvar{pubescence_color_all_nsamples}}
\signalplot{pubescence_color_all_T}{pubescence color (first GWAS)}{\emph{T}}%
{Vertical dotted lines indicate the location of the \getvar{pubescence_color_all_T_gene} gene associated with the locus.}
\geneplot{pubescence_color_all_T}{pubescence color (first GWAS)}{\emph{T}}{\getvar{pubescence_color_all_T_gene} gene}
\kmerplot{pubescence_color_all_T}{\getvar{pubescence_color_all_T_gene} gene}{\emph{T}}{pubescence color (first GWAS)}
\ldplot{pubescence_color_all}{pubescence color (first GWAS)}{\getvar{pubescence_color_all_ldkmers}}

\manhattanplot{pubescence_color_nogray}{pubescence color (second GWAS)}{\getvar{pubescence_color_nogray_nsamples}}
\signalplot{pubescence_color_nogray_Td}{pubescence color (second GWAS)}{\emph{Td}}%
{Vertical dotted lines indicate the location of the \getvar{pubescence_color_nogray_Td_gene} gene associated with the locus.}
\geneplot{pubescence_color_nogray_Td}{pubescence color (second GWAS)}{\emph{Td}}{\getvar{pubescence_color_nogray_Td_gene} gene}
\kmerplot{pubescence_color_nogray_Td}{\getvar{pubescence_color_nogray_Td_gene} gene}{\emph{Td}}{pubescence color (second GWAS)}

% Figures for seed coat color
\manhattanplot{seed_coat_color_all}{seed coat color (first GWAS)}{\getvar{seed_coat_color_all_nsamples}}
\signalplot{seed_coat_color_all_I}{seed coat color (first GWAS)}{\emph{I}}%
{Vertical dotted lines indicate the boundaries of the tandem duplication/inversion identified as the causal variant at this locus.}

\manhattanplot{seed_coat_color_greenyellow}{seed coat color (second GWAS)}{\getvar{seed_coat_color_greenyellow_nsamples}}
\signalplot{seed_coat_color_greenyellow_G}{seed coat color (second GWAS)}{\emph{G}}%
{Vertical dotted lines indicate the location of the \getvar{seed_coat_color_greenyellow_G_gene} gene associated with the locus.}
\geneplot{seed_coat_color_greenyellow_G}{seed coat color (second GWAS)}{\emph{G}}{\getvar{seed_coat_color_greenyellow_G_gene} gene}
\kmerplot{seed_coat_color_greenyellow_G}{\getvar{seed_coat_color_greenyellow_G_gene} gene}{\emph{G}}{seed coat color (second GWAS)}

% Figures for stem termination
\manhattanplot{stem_termination_all}{stem termination type (first GWAS)}{\getvar{stem_termination_all_nsamples}}
\signalplot{stem_termination_all_Dt1}{stem termination type (first GWAS)}{\emph{Dt1}}%
{Vertical dotted lines indicate the location of the \getvar{stem_termination_all_Dt1_gene} gene associated with the locus.}
\geneplot{stem_termination_all_Dt1}{stem termination type (first GWAS)}{\emph{Dt1}}{\getvar{stem_termination_all_Dt1_gene} gene}
\kmerplot{stem_termination_all_Dt1}{\getvar{stem_termination_all_Dt1_gene} gene}{\emph{Dt1}}{stem termination type (first GWAS)}

\manhattanplot{stem_termination_sn}{stem termination type (second GWAS)}{\getvar{stem_termination_sn_nsamples}}
\ldplot{stem_termination_sn}{stem termination type (second GWAS)}{\getvar{stem_termination_sn_ldkmers}}

% Figures for hilum color
\manhattanplot{hilum_color_all}{hilum color (first GWAS)}{\getvar{hilum_color_all_nsamples}}
\signalplot{hilum_color_all_T}{hilum color (first GWAS)}{\emph{T}}%
{Vertical dotted lines indicate the location of the \getvar{hilum_color_all_T_gene} gene associated with the locus.}
\geneplot{hilum_color_all_T}{hilum color (first GWAS)}{\emph{T}}{\getvar{hilum_color_all_T_gene} gene}
\kmerplot{hilum_color_all_T}{\getvar{hilum_color_all_T_gene} gene}{\emph{T}}{hilum color (first GWAS)}
\signalplot{hilum_color_all_I}{hilum color (first GWAS)}{\emph{I}}%
{Vertical dotted lines indicate the boundaries of the tandem duplication/inversion identified as the causal variant at this locus.}
\geneplot{hilum_color_all_I}{hilum color (first GWAS)}{\emph{I}}{causal tandem duplication/inversion}

\manhattanplot{hilum_color_blackbrown}{hilum color (second GWAS)}{\getvar{hilum_color_blackbrown_nsamples}}
\signalplot{hilum_color_blackbrown_R}{hilum color (second GWAS)}{\emph{R}}%
{Vertical dotted lines indicate the location of the \getvar{hilum_color_blackbrown_R_gene} gene associated with the locus.}
\geneplot{hilum_color_blackbrown_R}{hilum color (second GWAS)}{\emph{R}}{\getvar{hilum_color_blackbrown_R_gene}}
\kmerplot{hilum_color_blackbrown_R}{\getvar{hilum_color_blackbrown_R_gene} gene}{\emph{R}}{hilum color (second GWAS)}

\manhattanplot{hilum_color_rbr}{hilum color (third GWAS)}{\getvar{hilum_color_rbr_nsamples}}

% Figures for pod color
\manhattanplot{pod_color_all}{pod color (first GWAS)}{\getvar{pod_color_all_nsamples}}
\signalplot{pod_color_all_L1}{pod color (first GWAS)}{\emph{L1}}{}
\signalplot{pod_color_all_L2}{pod color (first GWAS)}{\emph{L2}}{}

\manhattanplot{pod_color_blbr}{pod color (second GWAS)}{\getvar{pod_color_blbr_nsamples}}
\ldplot{pod_color_blbr}{pod color (second GWAS)}{\getvar{pod_color_blbr_ldkmers}}
\signalplot{pod_color_blbr_pdcGm15}{pod color (second GWAS)}{newly suggested \emph{pdcGm15}}{}

% Figures for pubescence form
\manhattanplot{pubescence_form_all}{pubescence form (first GWAS)}{\getvar{pubescence_form_all_nsamples}}
\signalplot{pubescence_form_all_Pa1}{pubescence form (first GWAS)}{\emph{Pa1}}%
{Vertical dotted lines indicate the location of the \getvar{pubescence_form_all_Pa1_gene} gene that we suggest as a candidate for this locus.}
\geneplot{pubescence_form_all_Pa1}{pubescence form (first GWAS)}{\emph{Pa1}}{\getvar{pubescence_form_all_Pa1_gene} candidate gene putatively}
\kmerplot{pubescence_form_all_Pa1}{\getvar{pubescence_form_all_Pa1_gene} candidate gene putatively}{\emph{Pa1}}{pubescence form (first GWAS)}
\signalplot{pubescence_form_all_Pa2}{pubescence form (first GWAS)}{\emph{Pa2}}{}
\ldplot{pubescence_form_all}{pubescence form (first GWAS)}{\getvar{pubescence_form_all_ldkmers}}

\manhattanplot{pubescence_form_noerect}{pubescence form (second GWAS)}{\getvar{pubescence_form_noerect_nsamples}}

% Figures for pubescence density
\manhattanplot{pubescence_density}{pubescence density}{\getvar{pubescence_density_nsamples}}
\signalplot{pubescence_density_Ps}{pubescence density}{\emph{Ps}}%
{Vertical dotted lines indicate the boundaries of the causal CNV overlapping the \getvar{pubescence_density_Ps_gene} gene associated with the locus.}
\geneplot{pubescence_density_Ps}{pubescence density}{\emph{Ps}}{location of the causal CNV overlapping the \getvar{pubescence_density_Ps_gene} gene}

% Figures for seed coat luster
\manhattanplot{seed_coat_luster_all}{seed coat luster (first GWAS)}{\getvar{seed_coat_luster_all_nsamples}}

\manhattanplot{seed_coat_luster_nointermediate}{seed coat luster (second GWAS)}{\getvar{seed_coat_luster_nointermediate_nsamples}}

\manhattanplot{seed_coat_luster_dullshiny}{seed coat luster (third GWAS)}{\getvar{seed_coat_luster_dullshiny_nsamples}}
\ldplot{seed_coat_luster_dullshiny}{seed coat luster (third GWAS)}{\getvar{seed_coat_luster_dullshiny_ldkmers}}
\signalplot{seed_coat_luster_dullshiny_B}{seed coat luster (third GWAS)}{\emph{B}}%
{Vertical dotted lines indicate the boundaries of the causal CNV associated with the locus.}
\geneplot{seed_coat_luster_dullshiny_B}{seed coat luster (third GWAS)}{\emph{B}}{location of the causal CNV}

% Figures for corrected dry weight
\manhattanplot{corrected_dry_weight}{corrected dry weight (resistance to \emph{P. sojae}}{\getvar{corrected_dry_weight_nsamples}}
\ldplot{corrected_dry_weight}{corrected dry weight (resistance to \emph{P. sojae}}{\getvar{corrected_dry_weight_ldkmers}}
\signalplot{corrected_dry_weight_cdwGm15}{corrected dry weight (resistance to \emph{P. sojae}}{\emph{cdwGm15}}%
{Vertical dotted lines indicate the location of the \getvar{corrected_dry_weight_cdwGm15_gene} gene suggested by de Ronne et al. (2022) as associated with the locus.}

% Figures for maturity group
\manhattanplot{maturity_group}{maturity group}{\getvar{maturity_group_nsamples}}

% Figures for oil
\manhattanplot{oil}{seed oil content}{\getvar{oil_nsamples}}

% Figures for protein
\manhattanplot{protein}{seed protein content}{\getvar{protein_nsamples}}

\end{document}

